larq/quantizers.py

```diff
@@ -601,9 +604,95 @@ def get_config(self):
         return {**super().get_config(), "k_bit": self.precision}
 
 
+@utils.register_alias("dorefa_kernel_quantizer")
+@utils.register_keras_custom_object
+class DoReFaKernel(DoReFa):
+    r"""Instantiates a serializable k_bit kernel quantizer as in the DoReFa paper.
+
+    This quantizer is the same like the `DoReFa` quantizer, but adds a preprocessing.
+    Instead of limiting input operands (or in this case: weights) using a hard
+    limiter, a tangens hyperbolicus is applied to achieve a softer limiting
+    with a gradient, which is contineously differentiable itself.
+
+    \\[
+    w_{lim}(w) = tanh(w)
+    \\]
+
+    Furthermore, the weights of each layer are normed, such that the weight with
+    the largest magnitude gets the largest or smallest (depending on its sign)
+    quantizable value. That way, the full quantizable numeric range is utilized.
+
+    \\[
+    w_{norm}(w) = \frac{w}{max(|w|)}
+    \\]
+
+    The formulas can be found in the paper in section 2.3. Please note, that
+    the paper refers to weights being quantized on a numeric range of [-1,1], while
+    activations are quantized on the numeric range [0,1]. `DoReFa` defines the
+    quantization function quantizek() from the paper with the correct numeric
+    range of [0,1], which is why the range needs to be adapted before applying
+    `DoReFa` after applying the preprocessing (limit and norm).
+    The hard limiting inside `DoReFa` becomes ineffective, because its input
+    is already limited by the hyperbolic tangent. The full quantization
+    function including the adaption of numeric ranges is
+
+    \\[
+    q(w) = 2quantize_{k}(\frac{w_{norm}\left(w_{lim}\left(w\right)\right)}{2} + \frac{1}{2}) - 1
+    \\]
+
+    !!! warning
+        This quantizer works for weights on the range [-1,1], which matches the
+        default setting of `constraints.weight_clip`. Do not use this quantizer
+        with a different constraint *clip_value* than the default one.
+
+    ```plot-activation
+    quantizers.DoReFaKernel
+    ```
+
+    The interface of this quantizer is the same like the one of `DoReFa`.
+
+    # References
+        - [DoReFa-Net: Training Low Bitwidth Convolutional Neural Networks with Low
+            Bitwidth Gradients](https://arxiv.org/abs/1606.06160)
+    """
+
+    def call(self, inputs):
+
+        limited = tf.math.tanh(inputs)
+
+        # Divider for max-value norm.
+        # If all elements are 0., we would get a div by zero.
+        # So when all elements are zero, nothing is normed. This is achieved by
+        # dividing by 1.
+        dividend = tf.math.reduce_max(tf.math.abs(limited))
```

Comment:
```suggestion
        dividend = 2 * tf.math.reduce_max(tf.math.abs(limited))
```

Reply:
Sorry when I resolved https://github.com/larq/larq/pull/651#discussion_r620329710, I did not see that you include the scaling factor into the dividend. I suggest keeping all the scaling in the last two code lines, to have the scaling and the reversal scaling all in one place. See 426c64e93f029d53b8239774e262dc1ea9bb509c .